feat: Use `LambdaWorks`' `FieldElement` in `stark-felt` crate

fmoletta opened the pull request:
This PR implements the current api in the stark-felt crate using the [LambdaWorks](https://github.com/lambdaclass/lambdaworks) library

LucasLvy commented on crates/stark-felt/src/arbitrary.rs:
```diff
@@ -0,0 +1,48 @@
+const FIELD_HIGH: u128 = (1 << 123) + (17 << 64); // this is equal to 10633823966279327296825105735305134080
+const FIELD_LOW: u128 = 1;
+
+/// Returns a [`Strategy`] that generates any valid Felt
```
Can you elaborate more on when we should use this function ? more generally when do we need a `Strategy`.

LucasLvy commented on crates/stark-felt/src/arbitrary.rs:
```diff
@@ -0,0 +1,48 @@
+        .prop_map(Felt)
+}
+
+/// Returns a [`Strategy`] that generates any nonzero Felt
```
Same as above

LucasLvy commented on crates/stark-felt/src/lib.rs:
```diff
@@ -26,140 +37,193 @@ pub struct FromBytesError;
     /// Creates a new [Felt] from its big-endian representation in a [u8] slice.
-    pub const fn from_bytes_be(_bytes: &[u8]) -> Result<Self, FromBytesError> {
-        todo!()
+    pub fn from_bytes_be(bytes: &[u8]) -> Result<Self, FromBytesError> {
```
can you add in the rust doc how fast is this function compared to the other ways of creating a felt.

LucasLvy commented on crates/stark-felt/src/lib.rs:
```diff
@@ -26,140 +37,193 @@ pub struct FromBytesError;
     /// Creates a new [Felt] from its little-endian representation in a [u8] slice.
-    pub const fn from_bytes_le(_bytes: &[u8]) -> Result<Self, FromBytesError> {
-        todo!()
+    pub fn from_bytes_le(bytes: &[u8]) -> Result<Self, FromBytesError> {
```
can you add in the rust doc how fast is this function compared to the other ways of creating a felt.

LucasLvy commented on crates/stark-felt/src/lib.rs:
```diff
@@ -26,140 +37,193 @@ pub struct FromBytesError;
     /// Converts to big-endian byte representation in a [u8] array.
-    pub const fn to_bytes_be(&self) -> [u8; 32] {
-        todo!()
+    pub fn to_bytes_be(&self) -> [u8; 32] {
```
can you add in the rust doc how fast is this function

LucasLvy commented on crates/stark-felt/src/lib.rs:
```diff
@@ -26,140 +37,193 @@ pub struct FromBytesError;
     /// Converts to little-endian byte representation in a [u8] array.
-    pub const fn to_bytes_le(&self) -> [u8; 32] {
-        todo!()
+    pub fn to_bytes_le(&self) -> [u8; 32] {
```
can you add in the rust doc how fast is this function

LucasLvy commented on crates/stark-felt/src/lib.rs:
```diff
@@ -26,140 +37,193 @@ pub struct FromBytesError;
     /// Converts to big-endian bit representation.
-    pub const fn to_bits_be(&self) -> BitArray<BitArrayStore> {
-        todo!()
+    pub fn to_bits_be(&self) -> BitArray<BitArrayStore> {
```
can you add in the rust doc how fast is this function

LucasLvy commented on crates/stark-felt/src/lib.rs:
```diff
@@ -26,140 +37,193 @@ pub struct FromBytesError;
     /// Converts to little-endian bit representation.
-    pub const fn to_bits_le(&self) -> BitArray<BitArrayStore> {
-        todo!()
+    pub fn to_bits_le(&self) -> BitArray<BitArrayStore> {
```
can you add in the rust doc how fast is this function

LucasLvy reviewed (changes requested):
Can you make sure that it's no std compatible (i see some std imports).
More generally this crate aims to be used by everyone so can you please guide the devs to use the most performant functions plz (if they are all as performant as each other can you state it also).

fmoletta commented on crates/stark-felt/src/lib.rs:
```diff
@@ -26,140 +37,193 @@ pub struct FromBytesError;
-        todo!()
+    pub fn sqrt(&self) -> Option<Self> {
+        let (root_1, root_2) = self.0.sqrt()?;
+        Some(Self(if root_1 < root_2 { root_1 } else { root_2 }))
```
Sounds good!

fmoletta commented:
> Can you make sure that it's no std compatible (i see some std imports).
> More generally this crate aims to be used by everyone so can you please guide the devs to use the most performant functions plz (if they are all as performant as each other can you state it also).

Hello! Thanks for the review!
Regarding no-std: I found some imports were missing for no-std so I added them in (41d8e81), and made sure I could compile a no-std project importing stark-felt. I couldn't find any explicit std-imports not gated behind the "std" feature, could you expand on this?
Regarding the comments on the `arbitrary` module: I added some small comments indicating that they are used to generate values for proptests. Is this clear enough or should I add examples / a link to the proptest documentation? This is also a private module, so end users shouldn't have to worry about it.
Regarding performance of certain methods: I ran some micro-benchmarks and wasn't able to find any substancial difference between them. I think it's a good idea to add benchmarks to the crate, so end users can gauge the performance themselves instead of relying on the documentation

fmoletta commented on crates/stark-felt/src/lib.rs:
```diff
@@ -26,140 +37,193 @@ pub struct FromBytesError;
+        let mut limbs = self.0.representative().limbs;
+        limbs.reverse();
+
+        #[cfg(not(target_pointer_width = "64"))]
```
(264bd77)

Mirko-von-Leipzig commented:
> Regarding performance of certain methods: I ran some micro-benchmarks and wasn't able to find any substancial difference between them. I think it's a good idea to add benchmarks to the crate, so end users can gauge the performance themselves instead of relying on the documentation

I think the comment was more referring to which of those representations is actually the same as the inner type. i.e. does the field element store big endian or little endian? Or is it some other representation internally. While this does "leak" the inner workings a bit, it might be meaningful for those chasing to save instructions.

For example, in pathfinder we store felts as big endian in the database only because our field element uses big endian under the hood so its a no-op to convert.

Whether or not this actually matters is of course debatable :)

spapinistarkware commented on crates/stark-felt/src/lib.rs:
```diff
@@ -382,22 +486,55 @@ mod formatting {
 
     /// Represents [Felt] in decimal by default.
     impl fmt::Display for Felt {
```
Should we display decimal or hexadecimal here?

fmoletta replied:
We have `LowerHex` & `UpperHex` implementations so it makes sense to display decimal here imo

spapinistarkware commented on crates/stark-felt/src/lib.rs:
```diff
@@ -382,22 +486,55 @@ mod formatting {
+                let digit = if current < ten {
+                    current.limbs[3] as u8
+                } else {
+                    (current.div_rem(&ten).1).limbs[3] as u8
```
Why div_rem twice?

fmoletta replied:
Fixed (2799d21)

fmoletta commented on crates/stark-felt/src/lib.rs:
```diff
@@ -26,140 +44,201 @@ pub struct FromBytesError;
-        todo!()
+    pub fn mul_mod(&self, rhs: &Self, p: &Self) -> Self {
+        Self(FieldElement::from(
+            &(self.0 * rhs.0)
```
@xJonathanLEI has a use case for it in [starknet-rs](https://github.com/xJonathanLEI/starknet-rs/blob/ed14f483bb5d5171e6775960a24ee07018f3ff7c/starknet-crypto/src/ecdsa.rs#L88-L118).
I left it as it is part of the initial api, but we could discuss removing it and adding a `mod` method that specifically states that it performs modulo a number smaller than prime to support those use cases. This would also remove the need for `inverse_mod`.

spapinistarkware replied:
I find this confusing and a potential trap. If noone opposed, I say remove.

fmoletta replied:
Removed (5fc8a5a)

fmoletta commented on crates/stark-felt/src/lib.rs:
```diff
@@ -325,52 +404,77 @@ mod arithmetic {
 
     use super::*;
 
     impl Serialize for Felt {
```
Fixed (61c553b)

fmoletta commented on crates/stark-felt/src/lib.rs:
```diff
@@ -26,140 +44,201 @@ pub struct FromBytesError;
+        #[cfg(not(target_pointer_width = "64"))]
+        // Split limbs to adjust to BitArrayStore = [u32; 8]
+        let limbs: [u32; 8] = limbs
+            .map(|n| [(n >> 32) as u32, n as u32])
```
Fixed (29df244)

fmoletta commented on crates/stark-felt/src/lib.rs:
```diff
@@ -325,52 +404,77 @@ mod arithmetic {
-        fn sum<I: Iterator<Item = Self>>(_iter: I) -> Self {
-            todo!()
+        fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
+            iter.fold(Self::ZERO, |augend, addend| augend + addend)
```
Fixed (148d086)

fmoletta commented on crates/stark-felt/src/lib.rs:
```diff
@@ -26,140 +44,201 @@ pub struct FromBytesError;
-        todo!()
+    /// This is as performant as [to_bits_le](Felt::to_bits_le)
+    pub fn to_bits_be(&self) -> BitArray<BitArrayStore> {
+        let mut limbs = self.0.representative().limbs;
```
@LucasLvy what do you think about this?

LucasLvy replied:
If it changes it should be updated but IDK if it could change so much except if you do a big refacto on the felt type ? @fmoletta

xJonathanLEI replied:
IMO even if it changes the worst that can happen is that downstream applications see a performance drop if they bump dependency and didn't notice (or just didn't lock dependencies). They still do compile. They would notice if they really care about performance anyways.

If we really want to, we can always make a semver-incompatible release so that Cargo doesn't auto upgrade.
